articles/clojure_elixir_fibonacci_mstalker/article.md

```diff
@@ -0,0 +1,339 @@
+
+The Fibonacci sequence looks like this:
+
+```
+0, 1, 1, 2, 3, 5, 8, 13, 21, ...
+```
+
+The first two elements are `0` and `1`. After that, each element is the sum of the previous two elements.
+
+Here's today's programming challenge:
+
+> For any non-negative integer `n`, write a function to retrieve the `n`th Fibonacci number. Use a zero-based index.
+
+For example, when `n` is `4`, our function should return `3`.
+
+My goal in the following code exploration is to highlight some language features of Clojure and Elixir. I'll give examples of how someone might solve our programming challenge in either language, and then I'll explain some syntax or conventions. I'm not attempting to produce the most efficient or the cleverest code, as that may actually get in the way of giving you a taste for each language!
+
+
+### Clojure
+
+A straightforward solution to our problem looks like this in Clojure:
+
+```clojure
+(ns fibonacci)
+
+(defn compute
+  [n]
+  (cond
+    (= n 0) 0
+    (= n 1) 1
+    :else (+ (compute (- n 1))
+             (compute (- n 2)))))
+```
+
+We created a `fibonacci` namespace on the first line, and then we used `defn` to define a `compute` function. The next line, `[n]`, indicates that the function has one argument, which we call `n`.
+
+You'll notice all the parentheses. A pair of parentheses creates a list. The first element of a list is generally a function or macro, and the rest of the elements are the arguments.
+
+`cond` begins a conditional. If `n` is equal to zero, the function returns `0`. If `n` is `1`, then `compute` returns `1`. Mathematical operators like `=`, `+`, and `-` are just functions in Clojure, and they follow the same syntactical rules. That's why these operators appear before their arguments, rather than between them.
+
+You might have observed that no commas separate function arguments. That's normal in Clojure. Commas are optional in collections.
+
+Finally, we use the `:else` keyword to do something if `n` isn't `0` or `1`. We recursively call the `compute` function twice, passing in different arguments. This sums the two previous Fibonacci values. Putting the two `compute` function calls on separate lines is aesthetic. I could have written this:
+
+```clojure
+:else (+ (compute (- n 1)) (compute (- n 2)))
+```
+
+Clojure has no explicit `return`, so functions just return the last value they evaluate.
+
+If we wanted to, we could shorten the implementation slightly by using an inequality:
+
+```clojure
+(ns fibonacci)
+
+(defn compute
+  [n]
+  (cond
+    (< n 2) n
+    :else (+ (compute (- n 1))
+             (compute (- n 2)))))
+```
+
+We can call our function like this:
+
+```clojure
+(fibonacci/compute 10)
+;; 55
+```
+
+A semicolon starts a comment in Clojure. I wrote the return value of our function call after `;;`. Using two semicolons is idiomatic, and the language would have allowed us to use just one.
+
```

Comment:
I found the above a bit confusing. If you want to demonstrate comments, I'd do it in the context of your function code. Maybe show the output of the function call in a terminal screen shot or something. Better yet, offer a "try it!" feature where people can interactively call the function in an embedded REPL.